alarm_active_table.cpp

```diff
@@ -0,0 +1,423 @@
+      // and hence this will always be the 0.0 OID
+      snmp_set_var_typed_value(var, ASN_OBJECT_ID,
+                               (u_char*) zero_dot_zero_oid,
+                               sizeof(zero_dot_zero_oid));
+    }
+    break;
+    
+    
+    default: /** We shouldn't get here */
+    {
+      TRC_ERROR("unknown column: alarmActiveTable_get_value");
+      return SNMP_ERR_GENERR;
+    }
+  }
+
+  return SNMP_ERR_NOERROR;
+}
+
+/************************************************************
+ * 
+ *  Create a new row context and initialize its oid index.
+ */
+void alarmActiveTable_create_row(char* name,
+                                 alarmActiveTable_SNMPDateTime* datetime,
+                                 unsigned long index,
+                                 AlarmTableDef& def)
+{
+  alarmActiveTable_context* ctx = SNMP_MALLOC_TYPEDEF(alarmActiveTable_context);
+  ctx->_alarm_table_def = &def;
+  if (pointers[ctx->_alarm_table_def->alarm_index()])
+  {
+    if (ctx->_alarm_table_def->severity() == pointers[ctx->_alarm_table_def->alarm_index()]->_alarm_table_def->severity())
+    {
+      return;
+    }
+  }
+  if (!ctx)
+  {
+    TRC_ERROR("malloc failed: alarmActiveTable_create_row");
+    return;
+  }
+        
+  if (alarmActiveTable_index_to_oid(name, datetime, index, &ctx->_index) != SNMP_ERR_NOERROR)
+  {
+    if (ctx->_index.oids != NULL)
+    {
+      free(ctx->_index.oids);
+    }
+
+    free(ctx);
+    return;
+  }
+  if (ctx->_alarm_table_def->severity() == 1)
+  {
+    alarmActiveTable_delete_row(def);
+  }
+  else
+  {
+    if (pointers[ctx->_alarm_table_def->alarm_index()])
+    {
+      CONTAINER_REMOVE(cb.container, pointers[ctx->_alarm_table_def->alarm_index()]);
+    }
+    
+    if (ctx)
+    {
+      CONTAINER_INSERT(cb.container, ctx);
+      pointers[ctx->_alarm_table_def->alarm_index()] = ctx;
+    }
+  }
+  return;
+}
+
+/************************************************************
+ *
+ * Delete a row
+ */
+void alarmActiveTable_delete_row(AlarmTableDef& def)
+{
+  alarmActiveTable_context* ctx = SNMP_MALLOC_TYPEDEF(alarmActiveTable_context);
+  ctx->_alarm_table_def = &def;
+  if (!ctx)
+  {
+    TRC_ERROR("malloc failed: alarmActiveTable_remove_row");
+    return;
+  }
+  
+  if (ctx)
+  {
+    CONTAINER_REMOVE(cb.container, pointers[ctx->_alarm_table_def->alarm_index()]);
+    pointers[ctx->_alarm_table_def->alarm_index()] = NULL;
+  }
+}
+
+/************************************************************
+ *
+ *  Convert table index components to an oid.
+ */
+int alarmActiveTable_index_to_oid(char* name,
+                                 alarmActiveTable_SNMPDateTime* datetime,
+                                 unsigned long counter,
+                                 netsnmp_index *oid_idx)
+{
+  int err = SNMP_ERR_NOERROR;
+
+
+  netsnmp_variable_list var_alarmListName;
+  netsnmp_variable_list var_alarmActiveDateAndTime;
+  netsnmp_variable_list var_alarmActiveIndex;
```

Comment:
You should be able to make this area more succinct by using [designated initializers](https://gcc.gnu.org/onlinedocs/gcc/Designated-Inits.html):

`netsnmp_variable_list var_alarmActiveIndex = {.type = ASN_UNSIGNED, .next_variable = NULL};`